### DISCUSSION TITLE
Helping admins to manage deprecations better

### DISCUSSION THREAD

**Discussion by ahus1** [keycloak/keycloak#34365] [Archive]
**TL;DR:** Do we need a warning in the admin UI about deprecated options, or do we need a metric to monitor your installation for deprecated options?
I think the warnings we have today are too simple to miss for a busy admin.

I've been to a devops camp recently, and I heard a story where someone mentioned that Keycloak was simple to upgrade but then one of the upgrades just failed. It seems that they missed to migrate away from a deprecated feature before doing the next upgrade, and then Keycloak didn't start due to the change in the CLI options.

They fully automated the upgrade (in their case using Watchtower which automatically updates containers in a Docker environment). A similar thing could IMHO also happen with the Keycloak Operator. I found that they missed noticing the deprecations in the release notes, and also missed the warnings in the logs. 

I assume this is quite normal if you are the admin of 10s or 100s of applications that you want to update automatically, especially if they run simple and standard Keycloak installations. So we should have better UX around that. 

With the deprecation strategy we have, it enables them to do one major upgrade, but then they need to migrate their options before we remove the old feature. With the plan to issue minor releases for a while this might seem less relevant at first, but then it is even more relevant as I would assume that each deprecated features would be replaced with a new and better feature, and they would miss out on that one. 

So I wonder how to make the need to move away from deprecated features more obvious. 

Possible options I see: 

* Warnings on the admin console (similar to the temporary admin), possibly with the need to snooze the warnings for a week or two.
* A metric name with a label for each deprecated feature, so you can monitor and alert for all deprecated features you use.  

Happy to hear your ideas!

**Comment by andymunro** [comment]
Upgrades tend to be one of the key concerns of users, so it would be good to make it more predictable.  Also, it would be helpful if the release notes had a separate section for deprecated and removed features rather than just an entry mixed into a list of new features.  Then the deprecated features would  be more noticeable.

**Comment by stianst** [comment]
In 26.x a feature could be replaced by a newer version (with breaking changes) or could be deprecated fully. In terms of taking action to those an admin would not need to do anything before upgrading to 27, but could of course choose to do those gradually over time.

I don't think alerts in the admin console is the correct way to go; as it would be something like "hey the feature you are using will potentially be removed in 27, but that's two years from now. Do you want to sort that out now? Or do you want to dismiss this alert for X months?".

I think the first step is to improve the docs around this; as what we have now right now on https://www.keycloak.org/server/features is not the best. As a second step we could look at how to get the status of a given deployment. I could think something like `kc.sh features --status` or whatever would be nice, but we would want something equivalent to those for instance using the Operator. There it depends a bit on the target audience and how they would want the information. Could be metrics, could be a REST API (with a screen in the admin console), could also be a field in the CR.

**Reply by ahus1** [reply]
We can word it in a way "The feature you are using is deprecated, and there is now a better alternative". 

I know in some organisations, you must not use deprecated features, so people would be upgrading. And if we wouldn't advocate for it, we would get the feedback on the new things only after KC27 is released. 

So let's consider such an alert "educating users".

**Comment by jonkoops** [comment]
I believe the UI team is already working on some kind of landing page dashboard, adding some warnings for users of deprecated features seems reasonable to me. @ssilvert WDYT?

**Reply by ssilvert** [reply]
> I believe the UI team is already working on some kind of landing page dashboard, adding some warnings for users of deprecated features seems reasonable to me. @ssilvert WDYT?

I don't see any harm in considering this when we do refinement on the dashboard requirements.  Added a [comment ](https://github.com/keycloak/keycloak/issues/33712#issuecomment-2441630001) to the GitHub issue for it.
